### ISSUE TITLE
A RuntimeError when reproduced code, ask for your help :)

### ISSUE THREAD

**Issue by HardWayShip** [Terminator8758/CAP-master#6]
Hello. When I reproduced your code, the following error occurred 

![2021-12-10 23-16-22 的屏幕截图](https://user-images.githubusercontent.com/92527626/145598609-a69b1628-c083-490f-b1ea-e83e05d9c4e8.png)

the error occurred in this line, so how to fix it? Looking forward to your answer. Thanks.
![2021-12-10 23-21-13 的屏幕截图](https://user-images.githubusercontent.com/92527626/145598656-8744f329-d1ca-4f02-b160-64fdef38734a.png)

**Comment by Terminator8758** [comment]
Hi, the runtime error appeared when calling the ExemplarMemory(). This is related to the Pytorch version you use. When Pytorch version is high (e.g. >=1.8.0), the current implementation of ExemplarMemory() may need to change to a static form like this:
```
class ExemplarMemory(autograd.Function):

    @staticmethod
    def forward(ctx, inputs, indexes, features, momentum):
        ctx.features = features
        ctx.momentum = momentum
        ctx.save_for_backward(inputs, indexes)
        outputs = inputs.mm(ctx.features.t())

        return outputs

    @staticmethod
    def backward(ctx, grad_outputs):
        inputs, indexes = ctx.saved_tensors
        grad_inputs = None
        if ctx.needs_input_grad[0]:
            grad_inputs = grad_outputs.mm(ctx.features)

        # momentum update
        for x, y in zip(inputs, indexes):
            ctx.features[y] = ctx.momentum * ctx.features[y] + (1. - ctx.momentum) * x
            ctx.features[y] /= ctx.features[y].norm()

        return grad_inputs, None, None, None
```

**Comment by hahapt** [comment]
> Hi, the runtime error appeared when calling the ExemplarMemory(). This is related to the Pytorch version you use. When Pytorch version is high (e.g. >=1.8.0), the current implementation of ExemplarMemory() may need to change to a static form like this:
> 
> ```
> class ExemplarMemory(autograd.Function):
> 
>     @staticmethod
>     def forward(ctx, inputs, indexes, features, momentum):
>         ctx.features = features
>         ctx.momentum = momentum
>         ctx.save_for_backward(inputs, indexes)
>         outputs = inputs.mm(ctx.features.t())
> 
>         return outputs
> 
>     @staticmethod
>     def backward(ctx, grad_outputs):
>         inputs, indexes = ctx.saved_tensors
>         grad_inputs = None
>         if ctx.needs_input_grad[0]:
>             grad_inputs = grad_outputs.mm(ctx.features)
> 
>         # momentum update
>         for x, y in zip(inputs, indexes):
>             ctx.features[y] = ctx.momentum * ctx.features[y] + (1. - ctx.momentum) * x
>             ctx.features[y] /= ctx.features[y].norm()
> 
>         return grad_inputs, None, None, None
> ```

What does the 'indexes' mean?
And how to calculate the percam_inputs by using ExemplarMemory?

**Comment by Terminator8758** [comment]
> > Hi, the runtime error appeared when calling the ExemplarMemory(). This is related to the Pytorch version you use. When Pytorch version is high (e.g. >=1.8.0), the current implementation of ExemplarMemory() may need to change to a static form like this:
> > ```
> > class ExemplarMemory(autograd.Function):
> > 
> >     @staticmethod
> >     def forward(ctx, inputs, indexes, features, momentum):
> >         ctx.features = features
> >         ctx.momentum = momentum
> >         ctx.save_for_backward(inputs, indexes)
> >         outputs = inputs.mm(ctx.features.t())
> > 
> >         return outputs
> > 
> >     @staticmethod
> >     def backward(ctx, grad_outputs):
> >         inputs, indexes = ctx.saved_tensors
> >         grad_inputs = None
> >         if ctx.needs_input_grad[0]:
> >             grad_inputs = grad_outputs.mm(ctx.features)
> > 
> >         # momentum update
> >         for x, y in zip(inputs, indexes):
> >             ctx.features[y] = ctx.momentum * ctx.features[y] + (1. - ctx.momentum) * x
> >             ctx.features[y] /= ctx.features[y].norm()
> > 
> >         return grad_inputs, None, None, None
> > ```
> 
> What does the 'indexes' mean? And how to calculate the percam_inputs by using ExemplarMemory?

indexes are the memory label of inputs. To calculate percam_inputs, you can replace line 94 in CamAwareMemory.py with the following:
```
percam_inputs = ExemplarMemory.apply(percam_feat, mapped_targets, self.percam_memory[cc], torch.tensor(self.alpha).to(torch.device('cuda')))
```
